Microsoft.Toolkit.Uwp.UI.Controls.Core/ImageEx/ImageExBase.Source.cs

```diff
@@ -165,64 +160,90 @@ private async Task LoadImageAsync(Uri imageUri)
+        /// <summary>
+        /// This method is provided in case a developer would like their own custom caching strategy for <see cref="ImageExBase"/>.
+        /// By default it uses the built-in UWP cache provided by <see cref="BitmapImage"/> and
+        /// the <see cref="Image"/> control itself. Call <see cref="AttachSource(ImageSource)"/> to set
+        /// the retrieved cache value to the image.
+        /// </summary>
+        /// <example>
+        /// <code>
+        /// try
+        /// {
+        ///     var propValues = new List&lt;KeyValuePair&lt;string, object>>();
+        ///
+        ///     if (DecodePixelHeight > 0)
+        ///     {
+        ///         propValues.Add(new KeyValuePair&lt;string, object>(nameof(DecodePixelHeight), D ecodePixelHeight));
+        ///     }
+        ///     if (DecodePixelWidth > 0)
+        ///     {
+        ///         propValues.Add(new KeyValuePair&lt;string, object>(nameof(DecodePixelWidth), D ecodePixelWidth));
+        ///     }
+        ///     if (propValues.Count > 0)
+        ///     {
+        ///         propValues.Add(new KeyValuePair&lt;string, object>(nameof(DecodePixelType), DecodePixelType));
+        ///     }
+        ///
+        ///     // A token could be provided here as well to cancel the request to the cache,
+        ///     // if a new image is requested. That token can be canceled in the OnNewSourceRequested method.
+        ///     var img = await ImageCache.Instance.GetFromCacheAsync(imageUri, true, initializerKeyValues: propValues);
+        ///
+        ///     lock (LockObj)
+        ///     {
+        ///         // If you have many imageEx in a virtualized ListView for instance
+        ///         // controls will be recycled and the uri will change while waiting for the previous one to load
+        ///         if (_currentSourceUri == imageUri)
+        ///         {
+        ///             AttachSource(img);
+        ///             ImageExOpened?.Invoke(this, new ImageExOpenedEventArgs());
+        ///             VisualStateManager.GoToState(this, LoadedState, true);
+        ///         }
+        ///     }
+        /// }
+        /// catch (OperationCanceledException)
+        /// {
+        ///     // nothing to do as cancellation has been requested.
+        /// }
+        /// catch (Exception e)
+        /// {
+        ///     lock (LockObj)
+        ///     {
+        ///         if (_currentSourceUri == imageUri)
+        ///         {
+        ///             ImageExFailed?.Invoke(this, new ImageExFailedEventArgs(e));
+        ///             VisualStateManager.GoToState(this, FailedState, true);
+        ///         }
+        ///     }
+        /// }
+        /// </code>
+        /// </example>
+        /// <param name="imageUri"><see cref="Uri"/> of the image to load from the cache.</param>
+        /// <returns><see cref="Task"/></returns>
+        protected virtual Task ProvideCachedResourceAsync(Uri imageUri)
```

Comment:
Should I rename this as `AttachCachedResourceAsync`? Or should I have it return the resource and call `AttachSource` in the parent method? Thoughts?

Reply:
I am leaning towards the override method returns the resource. I think it might be simpler for the consumer and for the implementation.

Reply:
Would it be something like 
```
protected virtual ImageSource GetSource(Uri image)
        {
            return new BitmapImage(image);

        }
```

Reply:
I guess that would put calling `AttachSource` into our hands, then would be have to handle the lock? Would that be a good thing, rather then let the consumer deal with it?